[WIP] Add Docker-compose based dev environment

Jean85 opened the pull request:
This is to try and fix #808. STILL WIP

[EDIT] Issuing a `docker-compose run --rm php bash` command gives you a CLI inside the PHP container, where it's possible to run composer and phpunit commands. More than one test is still failing.

stof commented on docker/php/Dockerfile:
```diff
@@ -0,0 +1,11 @@
+FROM php:5.6-cli
```
Travis runs on PHP 7.0, so I think that the prod runs on PHP 7 too, not on 5.6

Jean85 replied:
I'm using that since the composer.json says `"php": ">=5.6"`. Since this is OSS, it would mean a BC otherwise...

stof commented on docker/solr/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+   && apt-get install -y curl \
+   && apt-get clean
+
+ADD 
```
adding nothing ? Why ?

stof commented on docker/solr/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+COPY entrypoint.sh /usr/local/bin/solr-entrypoint.sh
+COPY *.xml /etc/solr/
+
+RUN chmod +x /usr/local/bin/solr-entrypoint.sh
```
is it needed ? couldn't you make the original file executable and have it preserved when copying ?

stof commented on docker/solr/entrypoint.sh:
```diff
@@ -0,0 +1,214 @@
+    echo "Started"
+}
+
+download_and_run() {
```
Downloading Solr could be done at build-time rather than at start time (and we don't need the support of multiple Solr versions, as Packagist wants to use only 3.6, and if it migrates to Solr 6 in the future, it should also migrate to the official image)

Jean85 replied:
Done!

stof commented on docker/solr/schema.xml:
```diff
@@ -1,282 +1,282 @@
-<?xml version="1.0" encoding="UTF-8" ?>
-<schema name="packagist" version="1.4">
```
Why are you changing all lines in these files ? Are you changing line endings ?

Jean85 replied:
It was my IDE, it changed line separators.

stof replied:
if the file has CRLF in it currently, normalizing to LF is fine (we should have LF)

Jean85 commented:
Thanks @stof for the review, I've addressed all your concerns, apart from the PHP version one: in that case, we should decide if we need to bump the version on `composer.json` or not...

stof commented on docker/solr/entrypoint.sh:
```diff
@@ -0,0 +1,213 @@
+    fi
+}
+
+download_and_run 3.6.0
```
you don't need `download_and_run` anymore, as you download at build time now. You only need `run` here.

stof commented on docker/solr/entrypoint.sh:
```diff
@@ -0,0 +1,213 @@
+        fi
+        if [ -z "${SOLR_DOCS}" ]
+        then
+            echo "SOLR_DOCS not defined, skipping initial indexing"
```
IMO, you don't need this `SOLR_DOCS` stuff at all, as Packagist does not use it.

Jean85 commented:
Ok I've simplified the entrypoint script to avoid any unuseful cruft.

stof commented on docker-compose.yml:
```diff
@@ -0,0 +1,24 @@
+    build:
+      context: docker/solr
+    environment:
+      SOLR_CONFS: '/etc/solr' 
```
This should be set in the Dockerfile IMO, as this path actually depends on the Dockerfile itself (and so overriding it does not make sense).
We should keep this knowledge in a single place.

stof commented on docker/solr/entrypoint.sh:
```diff
@@ -0,0 +1,94 @@
+      done
+    fi
+
+    # enable custom core
```
This docker-compose setyp does not use custom core AFAICT, so this is also unused stuff

stof commented on docker/solr/entrypoint.sh:
```diff
@@ -0,0 +1,94 @@
+	version='3.6.0'
+    dir_name="apache-solr-${version}"
+    dir_conf="conf/"
+    add_core $dir_name $dir_conf $SOLR_CORE "$SOLR_CONFS"
```
this initialization could also be done at build time, as it only depends on the config files, which are copied in the image at build time.

Jean85 commented:
Ok I've centralized anything else remained, it should be ok now.
Now we can focus on making the tests green.

Seldaek commented:
Please stop right here.. IMO this is a waste of time. I don't even know what you are trying to solve here, but I'm sure introducing docker is not the solution. More later when I have time, sorry for the bluntness but I don't want you to waste more time on it than you already did.

Jean85 commented:
I'm trying to work on #808. Basically the CI is not running on all tests, and there are multiple ones that are broken, as they are right now.

Docker is just a reproducible way to run tests locally.

Seldaek commented:
@Jean85 I saw all that but sprinkling docker onto something is not the ultimate solution to all problems :p I have personally no interest in working with it and won't maintain this so I rather not have it in as it'll get outdated and end up confusing people more than anything.

Moving on to #808.
